src/jdk.compiler/share/classes/com/sun/tools/javac/comp/ThisEscapeAnalyzer.java

```diff
@@ -0,0 +1,1475 @@
+                    this.privateOuter |= (tree.mods.flags & Flags.PRIVATE) != 0;
+
+                    // Recurse
+                    super.visitClassDef(tree);
+                } finally {
+                    this.currentClass = currentClassPrev;
+                    this.privateOuter = privateOuterPrev;
+                    this.lint = lintPrev;
+                }
+            }
+
+            @Override
+            public void visitVarDef(JCVariableDecl tree) {
+                final Lint lintPrev = this.lint;
+                this.lint = this.lint.augment(tree.sym);
+                try {
+                    // Track warning suppression
+                    if (!this.lint.isEnabled(Lint.LintCategory.THIS_ESCAPE))
+                        ThisEscapeAnalyzer.this.suppressed.add(tree.sym);
+
+                    // Recurse
+                    super.visitVarDef(tree);
+                } finally {
+                    this.lint = lintPrev;
+                }
+            }
+
+            @Override
+            public void visitMethodDef(JCMethodDecl tree) {
+                final Lint lintPrev = this.lint;
+                this.lint = this.lint.augment(tree.sym);
+                try {
+                    // Track warning suppression
+                    if (!this.lint.isEnabled(Lint.LintCategory.THIS_ESCAPE))
+                        ThisEscapeAnalyzer.this.suppressed.add(tree.sym);
+
+                    // Determine if this is a constructor we should analyze
+                    final boolean analyzable = this.currentClassIsExternallyExtendable() &&
+                        TreeInfo.isConstructor(tree) &&
+                        !tree.sym.isPrivate() &&
+                        !ThisEscapeAnalyzer.this.suppressed.contains(tree.sym);
+
+                    // Determine if this method is "invokable" in an analysis (can't be overridden)
+                    final boolean invokable = !this.currentClassIsExternallyExtendable() ||
+                        TreeInfo.isConstructor(tree) ||
+                        (tree.mods.flags & (Flags.STATIC | Flags.PRIVATE | Flags.FINAL)) != 0;
+
+                    // Add method or constructor to map
+                    final MethodInfo info = new MethodInfo(this.currentClass, tree, analyzable, invokable);
+                    ThisEscapeAnalyzer.this.methodMap.put(tree.sym, info);
+
+                    // Recurse
+                    super.visitMethodDef(tree);
+                } finally {
+                    this.lint = lintPrev;
+                }
+            }
+
+            // Determines if the current class could be extended in some external compilation unit
+            private boolean currentClassIsExternallyExtendable() {
+                return !this.currentClass.sym.isFinal() &&
+                  !(this.currentClass.sym.isSealed() && this.currentClass.permitting.isEmpty()) &&
+                  !(this.currentClass.sym.owner.kind == MTH) &&
+                  !this.privateOuter;
```

Comment:
Unless you explicitly declare a nested class `private`, it won't have the `ACC_PRIVATE` flag, even though it is "implicitly private" because it has a `private` enclosing class.

Example:
```
$ cat PrivateOuter.java 
public class PrivateOuter {
    private static class Inner1 {
        static class Inner2 {
        }
    }
}
$ javap -v PrivateOuter\$Inner1\$Inner2
Classfile /Users/archie/proj/jdk/flex-test/classes/PrivateOuter$Inner1$Inner2.class
  Last modified Jan 12, 2023; size 408 bytes
  SHA-256 checksum 51ba6d39a5e66df2a078761d6424acbea7a8e32b8451f6ca7d2af49889673b2c
  Compiled from "PrivateOuter.java"
class PrivateOuter$Inner1$Inner2
  minor version: 0
  major version: 63
  flags: (0x0020) ACC_SUPER
  this_class: #7                          // PrivateOuter$Inner1$Inner2
  super_class: #2                         // java/lang/Object
  ...
```

So we have to keep track of this "implicit privateness" manually, which is what the `privateOuter` flag is for.

Reply:
D'oh - missed the `|=` - so this keeps updating...